toolchain/driver/clang_runner.cpp

```diff
@@ -0,0 +1,152 @@
+  total_size += exe_name_.size() + args.size() + 1;
+  if (inject_v_flag) {
+    total_size += strlen("-v") + 1;
+  }
+
+  // Allocate one chunk of storage for the actual C-strings and a vector of
+  // pointers into the storage.
+  llvm::OwningArrayRef<char> cstr_arg_storage(total_size);
+  llvm::SmallVector<const char*, 64> cstr_args;
+  cstr_args.reserve(args.size() + inject_v_flag + 1);
+
+  for (ssize_t i = 0; llvm::StringRef arg : llvm::concat<const llvm::StringRef>(
+                          llvm::ArrayRef({exe_name_}), maybe_v_flag, args)) {
+    cstr_args.push_back(&cstr_arg_storage[i]);
+    memcpy(&cstr_arg_storage[i], arg.data(), arg.size());
+    i += arg.size();
+    cstr_arg_storage[i] = '\0';
+    ++i;
+  }
+  for (const char* cstr_arg : llvm::ArrayRef(cstr_args).drop_front()) {
+    CARBON_VLOG() << "    '" << cstr_arg << "'\n";
+  }
+
+  CARBON_VLOG() << "Preparing Clang driver...\n";
+
+  // Create the diagnostic options and parse arguments controlling them out of
+  // our arguments.
+  llvm::IntrusiveRefCntPtr<clang::DiagnosticOptions> diagnostic_options =
+      clang::CreateAndPopulateDiagOpts(cstr_args);
+
+  // TODO: We don't yet support serializing diagnostics the way the actual
+  // `clang` command line does. Unclear if we need to or not, but it would need
+  // a bit more logic here to set up chained consumers.
+  clang::TextDiagnosticPrinter diagnostic_client(llvm::errs(),
+                                                 diagnostic_options.get());
+
+  clang::DiagnosticsEngine diagnostics(
+      diagnostic_ids_, diagnostic_options.get(), &diagnostic_client,
+      /*ShouldOwnClient=*/false);
+  clang::ProcessWarningOptions(diagnostics, *diagnostic_options);
+
+  clang::driver::Driver driver(exe_path_, target_, diagnostics);
+
+  // TODO: Directly run in-process rather than using a subprocess. This is both
+  // more efficient and makes debugging (much) easier. Needs code like:
+  // driver.CC1Main = [](llvm::SmallVectorImpl<const char*>& argv) {};
+  std::unique_ptr<clang::driver::Compilation> compilation(
+      driver.BuildCompilation(cstr_args));
+  CARBON_CHECK(compilation) << "Should always successfully allocate!";
+  if (compilation->containsError()) {
+    // These should have been diagnosed by the driver.
+    return false;
+  }
+
+  CARBON_VLOG() << "Running Clang driver...\n";
+
+  llvm::SmallVector<std::pair<int, const clang::driver::Command*>>
+      failing_commands;
+  int result = driver.ExecuteCompilation(*compilation, failing_commands);
+
+  // Finish diagnosing any failures before we verbosely log the source of those
+  // failures.
+  diagnostic_client.finish();
+
+  CARBON_VLOG() << "Execution result code: " << result << "\n";
+  for (const auto& [command_result, failing_command] : failing_commands) {
+    CARBON_VLOG() << "Failing command '" << failing_command->getExecutable()
+                  << "' with code '" << command_result << "' was:\n";
+    if (verbose()) {
```

Comment:
FWIW, this seems to only be done in this one spot, it might be worth inlining to reduce how much needs to be read.

Reply:
Sure. This too was a holdover from an older factoring of everything with different stream setup.

Reply:
(not sure if changes aren't published, but not seeing edits here)